File: ShoppingList/src/main/java/org/openintents/shopping/automation/EditAutomationActivity.java
package org.openintents.shopping.automation;

import android.app.Activity;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Spinner;
import android.widget.TextView;

import org.openintents.intents.AutomationIntents;
import org.openintents.intents.ShoppingListIntents;
import org.openintents.shopping.LogConstants;
import org.openintents.shopping.R;
import org.openintents.shopping.library.provider.ShoppingContract;

public class EditAutomationActivity extends Activity {

    private static final String TAG = LogConstants.TAG;
    private static final boolean debug = false || LogConstants.debug;

    private static final int REQUEST_CODE_PICK_LIST = 1;

    private static final String BUNDLE_ACTION = "action";
    private static final String BUNDLE_LIST_URI = "list";

    private TextView mTextCommand;
    // TextView mTextSelectAction;
    // TextView mTextSelectCountdown;
    private Spinner mSpinnerAction;
    private Button mButtonOk;
    private Button mButtonCountdown;

    private String mDescriptionAction;
    private String mDescriptionShoppingList;

    private Uri mUri;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_edit_automation);

        mUri = null;
        mDescriptionShoppingList = "?";

        mSpinnerAction = (Spinner) findViewById(R.id.spinner_action);
        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(
                this, R.array.automation_actions,
                android.R.layout.simple_spinner_item);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        mSpinnerAction.setAdapter(adapter);

        mSpinnerAction
                .setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {

                    @Override
                    public void onItemSelected(AdapterView<?> parent,
                                               View view, int position, long id) {
                        mDescriptionAction = getResources().getStringArray(
                                R.array.automation_actions)[position];

                        updateTextViews();
                    }

                    @Override
                    public void onNothingSelected(AdapterView<?> parent) {

                    }

                });

        mButtonCountdown = (Button) findViewById(R.id.button_list);
        mButtonCountdown.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                pickList();
            }

        });

        mButtonOk = (Button) findViewById(R.id.button_ok);
        mButtonOk.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                doOk();
            }

        });

        mButtonOk.setEnabled(false);

        Button b = (Button) findViewById(R.id.button_cancel);
        b.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(View v) {
                doCancel();
            }

        });

        mTextCommand = (TextView) findViewById(R.id.command);
        // mTextSelectAction = (TextView) findViewById(R.id.select_action);
        // mTextSelectCountdown = (TextView)
        // findViewById(R.id.select_countdown);

        if (savedInstanceState != null) {
            if (savedInstanceState.containsKey(BUNDLE_ACTION)) {
                int i = savedInstanceState.getInt(BUNDLE_ACTION);
                mSpinnerAction.setSelection(i);
            }
            if (savedInstanceState.containsKey(BUNDLE_LIST_URI)) {
                mUri = Uri.parse(savedInstanceState.getString(BUNDLE_LIST_URI));
                setListNameFromUri();
            }
        } else {
            final Intent intent = getIntent();

            if (intent != null) {
                String action = intent
                        .getStringExtra(ShoppingListIntents.EXTRA_ACTION);

                if (ShoppingListIntents.TASK_CLEAN_UP_LIST.equals(action)) {
                    mSpinnerAction.setSelection(0);
                } else {
                    // set default
                    mSpinnerAction.setSelection(0);
                }

                // Get list:

                final String dataString = intent
                        .getStringExtra(ShoppingListIntents.EXTRA_DATA);
                if (dataString != null) {
                    mUri = Uri.parse(dataString);
                }
                setListNameFromUri();
            }
        }

        updateTextViews();
    }

    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (debug) {
            Log.i(TAG, "onPause");
        }

    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt(BUNDLE_ACTION, mSpinnerAction.getSelectedItemPosition());
        if (mUri != null) {
            outState.putString(BUNDLE_LIST_URI, mUri.toString());
        }
    }

    void pickList() {
        Intent i = new Intent(Intent.ACTION_PICK);
        i.setData(ShoppingContract.Lists.CONTENT_URI);

        startActivityForResult(i, REQUEST_CODE_PICK_LIST);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode,
                                    Intent intent) {
        super.onActivityResult(requestCode, resultCode, intent);

        if (requestCode == REQUEST_CODE_PICK_LIST && resultCode == RESULT_OK) {
            mUri = intent.getData();

            setListNameFromUri();
        }

        updateTextViews();
    }

    private void setListNameFromUri() {
        mDescriptionShoppingList = "";

        if (mUri != null) {
            mButtonOk.setEnabled(true);

            // Get name of list from content provider
            Cursor c = getContentResolver().query(
                    mUri,
                    new String[]{ShoppingContract.Lists._ID,
                            ShoppingContract.Lists.NAME}, null, null, null
            );

            if (c != null && c.moveToFirst()) {
                mDescriptionShoppingList = c.getString(1);
            }

            c.close();
        }

        if (TextUtils.isEmpty(mDescriptionShoppingList)) {
            mDescriptionShoppingList = getString(android.R.string.untitled);
        }
    }

    void doOk() {
        updateResult();
        finish();
    }

    void doCancel() {
        setResult(RESULT_CANCELED);
        finish();
    }

    void updateResult() {
        Intent intent = new Intent();

        long id = mSpinnerAction.getSelectedItemId();
        if (id == 0) {
            intent.putExtra(ShoppingListIntents.EXTRA_ACTION,
                    ShoppingListIntents.TASK_CLEAN_UP_LIST);
        }
        intent.putExtra(ShoppingListIntents.EXTRA_DATA, mUri.toString());

        String description = mDescriptionAction + ": "
                + mDescriptionShoppingList;
        intent.putExtra(AutomationIntents.EXTRA_DESCRIPTION, description);

        if (debug) {
            Log.i(TAG, "Created intent (URI)   : " + intent.toURI());
        }
        if (debug) {
            Log.i(TAG, "Created intent (String): " + intent.toString());
        }

        setResult(RESULT_OK, intent);
    }

    void updateTextViews() {
        mTextCommand.setText(mDescriptionAction + ": "
                + mDescriptionShoppingList);
        // mTextSelectAction.setText(getString(R.string.select_action,
        // mDescriptionAction));
        // mTextSelectCountdown.setText(getString(R.string.select_countdown,
        // mDescriptionCountdown));
        // mTextSelectAction.setText(getString(R.string.select_action, ""));
        // mTextSelectCountdown.setText(getString(R.string.select_countdown,
        // ""));
        mButtonCountdown.setText(mDescriptionShoppingList);
    }
}

File: ShoppingList/src/main/java/org/openintents/shopping/automation/EditAutomationActivity.kt
package org.openintents.shopping.automation

import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.Spinner
import android.widget.TextView
import org.openintents.intents.AutomationIntents
import org.openintents.intents.ShoppingListIntents
import org.openintents.shopping.LogConstants
import org.openintents.shopping.R
import org.openintents.shopping.library.provider.ShoppingContract

open class EditAutomationActivity : Activity() {

    private lateinit var mTextCommand: TextView
    // private lateinit var mTextSelectAction: TextView
    // private lateinit var mTextSelectCountdown: TextView
    private lateinit var mSpinnerAction: Spinner
    private lateinit var mButtonOk: Button
    private lateinit var mButtonCountdown: Button

    private var mDescriptionAction: String? = null
    private var mDescriptionShoppingList: String? = null

    private var mUri: Uri? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_edit_automation)

        mUri = null
        mDescriptionShoppingList = "?"

        mSpinnerAction = findViewById<Spinner>(R.id.spinner_action)
        val adapter = ArrayAdapter.createFromResource(
            this, R.array.automation_actions,
            android.R.layout.simple_spinner_item
        )
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        mSpinnerAction.adapter = adapter

        mSpinnerAction.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {

            override fun onItemSelected(
                parent: AdapterView<*>,
                view: View?,
                position: Int,
                id: Long
            ) {
                mDescriptionAction = resources.getStringArray(
                    R.array.automation_actions
                )[position]

                updateTextViews()
            }

            override fun onNothingSelected(parent: AdapterView<*>) {
            }
        }

        mButtonCountdown = findViewById<Button>(R.id.button_list)
        mButtonCountdown.setOnClickListener {
            pickList()
        }

        mButtonOk = findViewById<Button>(R.id.button_ok)
        mButtonOk.setOnClickListener {
            doOk()
        }

        mButtonOk.isEnabled = false

        val b = findViewById<Button>(R.id.button_cancel)
        b.setOnClickListener {
            doCancel()
        }

        mTextCommand = findViewById<TextView>(R.id.command)
        // mTextSelectAction = findViewById(R.id.select_action)
        // mTextSelectCountdown = findViewById(R.id.select_countdown)

        if (savedInstanceState != null) {
            if (savedInstanceState.containsKey(BUNDLE_ACTION)) {
                val i = savedInstanceState.getInt(BUNDLE_ACTION)
                mSpinnerAction.setSelection(i)
            }
            if (savedInstanceState.containsKey(BUNDLE_LIST_URI)) {
                mUri = Uri.parse(savedInstanceState.getString(BUNDLE_LIST_URI))
                setListNameFromUri()
            }
        } else {
            val intent = intent

            if (intent != null) {
                val action = intent.getStringExtra(ShoppingListIntents.EXTRA_ACTION)

                if (ShoppingListIntents.TASK_CLEAN_UP_LIST == action) {
                    mSpinnerAction.setSelection(0)
                } else {
                    // set default
                    mSpinnerAction.setSelection(0)
                }

                // Get list:
                val dataString = intent.getStringExtra(ShoppingListIntents.EXTRA_DATA)
                if (dataString != null) {
                    mUri = Uri.parse(dataString)
                }
                setListNameFromUri()
            }
        }

        updateTextViews()
    }

    override fun onResume() {
        super.onResume()
    }

    override fun onPause() {
        super.onPause()
        if (debug) {
            Log.i(TAG, "onPause")
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(BUNDLE_ACTION, mSpinnerAction.selectedItemPosition)
        if (mUri != null) {
            outState.putString(BUNDLE_LIST_URI, mUri.toString())
        }
    }

    internal fun pickList() {
        val i = Intent(Intent.ACTION_PICK)
        i.data = ShoppingContract.Lists.CONTENT_URI

        startActivityForResult(i, REQUEST_CODE_PICK_LIST)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
        super.onActivityResult(requestCode, resultCode, intent)

        if (requestCode == REQUEST_CODE_PICK_LIST && resultCode == RESULT_OK) {
            mUri = intent?.data

            setListNameFromUri()
        }

        updateTextViews()
    }

    private fun setListNameFromUri() {
        mDescriptionShoppingList = ""

        if (mUri != null) {
            mButtonOk.isEnabled = true

            // Get name of list from content provider
            val c = contentResolver.query(
                mUri!!,
                arrayOf(ShoppingContract.Lists._ID, ShoppingContract.Lists.NAME),
                null, null, null
            )

            if (c != null && c.moveToFirst()) {
                mDescriptionShoppingList = c.getString(1)
            }

            c?.close()
        }

        if (TextUtils.isEmpty(mDescriptionShoppingList)) {
            mDescriptionShoppingList = getString(android.R.string.untitled)
        }
    }

    internal fun doOk() {
        updateResult()
        finish()
    }

    internal fun doCancel() {
        setResult(RESULT_CANCELED)
        finish()
    }

    internal fun updateResult() {
        val intent = Intent()

        val id = mSpinnerAction.selectedItemId
        if (id == 0L) {
            intent.putExtra(
                ShoppingListIntents.EXTRA_ACTION,
                ShoppingListIntents.TASK_CLEAN_UP_LIST
            )
        }
        intent.putExtra(ShoppingListIntents.EXTRA_DATA, mUri.toString())

        val description = mDescriptionAction + ": " + mDescriptionShoppingList
        intent.putExtra(AutomationIntents.EXTRA_DESCRIPTION, description)

        if (debug) {
            Log.i(TAG, "Created intent (URI)   : " + intent.toURI())
        }
        if (debug) {
            Log.i(TAG, "Created intent (String): " + intent.toString())
        }

        setResult(RESULT_OK, intent)
    }

    internal fun updateTextViews() {
        mTextCommand.text = mDescriptionAction + ": " + mDescriptionShoppingList
        // mTextSelectAction.setText(getString(R.string.select_action,
        // mDescriptionAction))
        // mTextSelectCountdown.setText(getString(R.string.select_countdown,
        // mDescriptionCountdown))
        // mTextSelectAction.setText(getString(R.string.select_action, ""))
        // mTextSelectCountdown.setText(getString(R.string.select_countdown,
        // ""))
        mButtonCountdown.text = mDescriptionShoppingList
    }

    companion object {
        private val TAG = LogConstants.TAG
        private val debug = false || LogConstants.debug

        private const val REQUEST_CODE_PICK_LIST = 1

        private const val BUNDLE_ACTION = "action"
        private const val BUNDLE_LIST_URI = "list"
    }
}
